### ISSUE TITLE
Improve color items UX in faceted search

### ISSUE THREAD

**Issue by mparvazi** [PrestaShop/theme-refacto#164]
### Current design:

<img src="https://user-images.githubusercontent.com/85633460/150693098-0ed8597f-232d-49f1-8e6d-e9c31dd4d116.png" width=600>

### Expected:

<img src="https://user-images.githubusercontent.com/85633460/150693113-e6f92b23-6bb6-4b23-be6c-cf0f82f7f7d2.png" width=600>

**Comment by NeOMakinG** [comment]
@prestascott you need to validate it, I like it honestly

**Comment by NeOMakinG** [comment]
@mparvazi 

![image](https://user-images.githubusercontent.com/14963751/150742584-527c5612-82eb-473f-a3c4-cbc56c66ecb5.png)


FYI, there are mockups that you can find here https://www.figma.com/file/LfVl5leeSKcVUhSaYwhbtM/New-Theme?node-id=154%3A11172

**Comment by mparvazi** [comment]
It's obvious that input checkbox shouldn't be alongside of color-box, as current theming (it's boring!)

**So IF we remove input checkbox for colors:**
When a color is selected in faceted search, we should somehow show to user that color is selected now 

There is some options:
1. A check icon before or after of label (as i did)
2. Bolding or Coloring the font of color-label.
3. Boxing the color-label with border or background color. 

I wanted to use check-icon inside the color-box, but problem is that color-box uses dynamic color code, and we have to use invert() color for check-icon. I tested but I didn't like it.

<img src="https://user-images.githubusercontent.com/85633460/150743977-2c7df8d4-13c7-4ff0-9ec1-858ad9f122c5.png" width="600">

**Comment by prestascott** [comment]
Hey everyone,

Thanks @mparvazi for your request!

Among your 3 options, I would choose the check icon when the color is selected. And I would also add the color inside the checkbox (even if it isn't checked). 

The problem here is that we're currently using Material icons checkboxes everywhere in the theme so It will be inconsistent if the border size isn't the same for colors, right?

According to the [WCAG](https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html), the non-text contrast ratio must at least be 3:1. So to ensure this minimum contrast ratio requirement, here is what I suggest about the check icon colors:
- When the checkbox background color is considered as light, use `#363A41` for the check icon
- When the checkbox background color is considered as dark, use `#FFFFFF` for the check icon

----
And here are my design propositions: 

### First proposition
- Each checkbox has a border to cover issues identifying the white color checkbox
- Reduced checkbox border size to 1 px
- These changes allow the theme to be consistent on every page
![filters (3)](https://user-images.githubusercontent.com/54888056/150816846-7f31d0c7-5161-4697-9e6b-d175c4e9b373.png)

### Second proposition
- Each checkbox has a border for the same reason as the first proposition
- Reduced checkbox border size to 1 px
- The shape of the checkbox used for colors is a circle so it's easier to identify
![filters (5)](https://user-images.githubusercontent.com/54888056/150818364-7d60261d-fbed-4f91-858d-61ccafe3e4a0.png)

### Last proposition
- Only unselected checkbox has a border (like the current)
![filters (6)](https://user-images.githubusercontent.com/54888056/150819792-ca09fb53-ae15-41f2-ae93-a9a421a069b4.png)




Tell me wdyt?

**Comment by mparvazi** [comment]
@NeOMakinG 

The color of check-mark for box-color must be calculated dynamically, because colors will be used as inline style:

![Screenshot from 2022-01-24 23-48-59](https://user-images.githubusercontent.com/85633460/150859313-781683d7-6019-43f4-8bd0-cc58301ddf7a.png)

My solution for dynamically choosing check-mark color based on box-color background:

![Screenshot from 2022-01-24 23-59-51](https://user-images.githubusercontent.com/85633460/150859745-86895c21-28dd-48f4-a5ca-c42f6fa480a6.png)

**Result:**
Color of check-mark will be 'white' or 'black', based on box-color background blackness.

<img src="https://user-images.githubusercontent.com/85633460/150859872-0c00448d-8550-42e4-a75e-124e9b4ee951.png" width="300">


Just a test, needs more work.

**Comment by mparvazi** [comment]
@prestascott 

First proposition, IMO consistency is more important!

**Comment by SharakPL** [comment]
No. This is all wrong! What if you have colorful texture instead of simple color? How are you going to set the color for the checkmark on a black and white texture?

IMO best way to do that is:
1. Add label wherever possible.
2. If checkmark will be added, then it has to be OUTSIDE color box.

**Comment by NeOMakinG** [comment]
I agree with @SharakPL on this, @prestascott it's impossible to calculate which color you should use for the checkmark icon, because if you have something else than a simple color, you won't be able to calculate it

**Comment by mparvazi** [comment]
> I agree with @SharakPL on this, @prestascott it's impossible to calculate which color you should use for the checkmark icon, because if you have something else than a simple color, you won't be able to calculate it

For textures we can use a shadow or border for check mark.
Check mark for textures is always white with black border or shadow.

**Comment by mparvazi** [comment]
@SharakPL 

Please don't tell me Adidas has got wrong!

**Comment by mparvazi** [comment]
![Screenshot 2022-01-25 at 12-26-32 Mens Tops T-Shirts Nike com](https://user-images.githubusercontent.com/85633460/150944543-86d0a827-9bf1-4fd5-86e7-92abf33c59df.png)

**Comment by mparvazi** [comment]
![Screenshot 2022-01-25 at 12-28-06 Mens Tops T-Shirts Nike com](https://user-images.githubusercontent.com/85633460/150944805-b64b0415-62fd-48e6-b473-0ba19808b64c.png)

**Comment by prestascott** [comment]
Yep we can use gradients like Adidas, but you don't have the latest version I think because they added the label:
![image](https://user-images.githubusercontent.com/54888056/150944838-5f6fc875-f63b-4a39-bc8a-1ab0f9a6bd9c.png)

Nike has a contrast ratio issue, see the yellow. You can't put a white check on yellow background 😅

**Comment by mparvazi** [comment]
@prestascott 

I used black check-mark for yellow in my code:
This [code](https://github.com/PrestaShop/theme-refacto/issues/164#issuecomment-1020524027) is smart.

<img src="https://user-images.githubusercontent.com/85633460/150859872-0c00448d-8550-42e4-a75e-124e9b4ee951.png" width="300">

**Comment by mparvazi** [comment]
@SharakPL 

I prefer to wait for @prestascott new mockups for textures and possibly the final design.
You can create a sample for your ideas (as I did), it's easier for @prestascott to make a decision.

**Comment by SharakPL** [comment]
> @SharakPL If you have a black and white texture then you will select 2 checkboxes: the black and the white. This is what e-commerce websites do today.
> Don't worry, there will always be a label. Please see my mockups just above.

@prestascott I'm not sure I follow. Which 2 checkboxes? There's gonna be 1 checkbox marked with a texture (can be anything). Are you saying you'll have to set a list of colors for any given texture in BO?

**Comment by SharakPL** [comment]
> Nike has a contrast ratio issue, see the yellow. You can't put a white check on yellow background 😅

There's always going to be a problem with this, because there's no way to have a distinguishable checkmark on every texture. IMO active color filter should be accented differently.

**Comment by mparvazi** [comment]
Let me check to find a solution for textures.

**Comment by SharakPL** [comment]
> Let me check to find a solution for textures.

I'm sure you will but that's not the point. Any additional script is bad for performance and scripts for this are simply unnecessary, when all you need is just add an _active_ class to checked attribute and let css do the work.

**Comment by prestascott** [comment]
@SharakPL Can you please explain to us what is your definition of texture? Because to me, it doesn't mean color.
You can have a plain color: black, white, blue, red, etc. And you can use gradient color for multicolor option, but this still isn't a texture.

**Comment by SharakPL** [comment]
If you have an attribute of type _color or texture_ then you either set a color from color pallet or upload a texture file. Either of it will be visible on the filters, right?

**Comment by mparvazi** [comment]
We can use JavaScript to get average color of textures:
Then we can calculate check-mark color based on average color.

![Screenshot 2022-01-25 at 13-57-11 Find Average Color of image via JavaScript ](https://user-images.githubusercontent.com/85633460/150959800-9250624d-067b-4e3a-88d9-376da7c251a8.png)

![Screenshot 2022-01-25 at 13-56-30 Find Average Color of image via JavaScript ](https://user-images.githubusercontent.com/85633460/150959815-d2c016b7-aa5c-4a2d-a16b-2997e8a9972f.png)

![Screenshot 2022-01-25 at 13-56-12 Find Average Color of image via JavaScript ](https://user-images.githubusercontent.com/85633460/150959830-4abf54c5-2782-4707-8759-67f9fcec4fe2.png)

![Screenshot 2022-01-25 at 13-55-49 Find Average Color of image via JavaScript ](https://user-images.githubusercontent.com/85633460/150959854-e3599e0b-18bf-4c75-ac8d-522718042e68.png)

**Comment by mparvazi** [comment]
<img src="https://user-images.githubusercontent.com/85633460/150961019-37d5d9b6-9988-4451-a49a-33347f622a06.png" width="300">

**Comment by SharakPL** [comment]
First of all what about unused checkboxes on the left? Will there be 2 checkboxes for colors? What's the point?

![obraz](https://user-images.githubusercontent.com/5007456/150971150-b1a98b8a-a70f-46d7-9935-aeac0422201a.png)

Of course you can hide it but can you tell if color is checked without it? Not always:

![obraz](https://user-images.githubusercontent.com/5007456/150972681-eeb52ad7-8474-4f6a-9c00-1025ddfca646.png)

**Comment by mparvazi** [comment]
> First of all what about unused checkboxes on the left? Will there be 2 checkboxes for colors? What's the point?
> 
> ![obraz](https://user-images.githubusercontent.com/5007456/150971150-b1a98b8a-a70f-46d7-9935-aeac0422201a.png)
> 
> Of course you can hide it but can you tell if color is checked without it? Not always:
> 
> ![obraz](https://user-images.githubusercontent.com/5007456/150972681-eeb52ad7-8474-4f6a-9c00-1025ddfca646.png)

The input checkboxes will be hide but exists in html code.
Then we use this code for controlling the selected inputs:

```
input{
    .check-mark {
        display: none
    }
}

input:checked {
    .check-mark {
        display: block
    }
}
```

This the implementation that I did:

![150743977-2c7df8d4-13c7-4ff0-9ec1-858ad9f122c5](https://user-images.githubusercontent.com/85633460/150973584-14c8360b-d6c6-4d4f-9294-c634676c87e4.png)

**Comment by mparvazi** [comment]
We can use svg icon with stroke for textures:

<img src="https://user-images.githubusercontent.com/85633460/150980732-625cfac1-e1af-4c73-aea9-43060efaef9f.png" width="300">

This svg icon can be redesign to match with check mark icon.

**Comment by mparvazi** [comment]
Average color function:

```
function averageColor(imageElement) {
	var canvas = document.createElement('canvas'),
	context = canvas.getContext && canvas.getContext('2d'),
	imgData, width, height,
	length,
	rgb = { r: 0, g: 0, b: 0 },
	count = 0;

	height = canvas.height =
		imageElement.naturalHeight ||
		imageElement.offsetHeight ||
		imageElement.height;
	width = canvas.width =
		imageElement.naturalWidth ||
		imageElement.offsetWidth ||
		imageElement.width;

	context.drawImage(imageElement, 0, 0);

	imgData = context.getImageData(0, 0, width, height);

	length = imgData.data.length;

	for (var i = 0; i < length; i += 4) {
		rgb.r += imgData.data[i];
		rgb.g += imgData.data[i + 1];
		rgb.b += imgData.data[i + 2];
		count++;
	}

	rgb.r = Math.floor(rgb.r / count);
	rgb.g = Math.floor(rgb.g / count);
	rgb.b = Math.floor(rgb.b / count);

	return rgb;
}
```

**Comment by mparvazi** [comment]
@NeOMakinG @prestascott 

## Texture test
1. Using svg icon (equal to icon-font used in theme)
I used [this](https://github.com/google/material-design-icons/tree/master/src/action/done/materialicons) icon for check-mark (google material icons)
```
<svg xmlns="http://www.w3.org/2000/svg" height="24" viewBox="0 0 24 24" width="24" class="check-mark">
  <path d="M0 0h24v24H0z" fill="none" class="check-mark-background"/>
  <path d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z" class="check-mark-icon"/>
</svg>
```

2. Getting the `average-color` of texture image
This can be implement by JavaScript
```
const boxTexture = filter.dataset.boxTexture;

let textureImg = new Image();
textureImg.src = boxTexture;
const checkMarkBaseColor = averageColor(textureImg);
```

3. Calculating check-mark color based on average-color blackness
```
const blackness = 100 - Math.max(Math.max(checkMarkBaseColor.r, checkMarkBaseColor.g), checkMarkBaseColor.b) / 255 * 100;
const checkMarkColor = (blackness > 25) ? 'white' : 'black';
checkMarkIcon.setAttribute('fill', checkMarkColor);
```

4. Using average-color for check-mark icon background with opacity 50%
```
.texture {
  svg {
    .check-mark-background {
      fill-opacity: 50%;
    }
  }
}
```

## Result

| Default | Filtered |
| ------------- | ------------- |
| <br><img src="https://user-images.githubusercontent.com/85633460/151015413-d2afe0fa-243f-471b-a3c6-7e4fa64a60b8.png" width="250">  | <br><img src="https://user-images.githubusercontent.com/85633460/151016798-1dbc92ed-81ed-4032-826a-c02f7f6ce88e.png" width="250">  |

**Comment by Hlavtox** [comment]
That opacity thing seems to work!

**Comment by NeOMakinG** [comment]
> That opacity thing seems to work!

I agree, looks fine

**Comment by prestascott** [comment]
Yep, the opacity that you added @mparvazi is working because there is a better contrast ratio with the check icon. Thanks!

**Comment by sciarasouani** [comment]
Good morning,
I would need to use the tick/icon inside the variant colors which are handled by the product-variants.tpl file, but I can't figure out how to do it.
Is there anyone who could help me? Thank you

**Comment by SharakPL** [comment]
@sciarasouani this is not a good place for that. Please join the community on [Slack](https://prestashop.slack.com/).
